### ISSUE TITLE
incompatible_disallow_empty_glob: fail if a glob doesn't match anything

### ISSUE THREAD

**Issue by laurentlb** [bazelbuild/bazel#8195]
The [`glob()`](https://docs.bazel.build/versions/master/be/functions.html#glob) function tends to be error-prone, because any typo in a path will silently return an empty list. This has been a source of confusion and bugs for many users.

`glob` has an optional Boolean argument `allow_empty`. For example, let's say your code looks like:
```python
glob([pattern_a, pattern_b], exclude = [pattern_c], allow_empty = False)
```

This code will fail if `pattern_a` or `pattern_b` doesn't match anything. It also fails if the whole function (after excluding `pattern_c`) doesn't match anything. We believe this behavior is safer in general.

In rare cases, this is intentional that a pattern doesn't match anything, e.g. when a source file is optional. When it happens, users should tell their intent explicitly by setting `allow_empty = True`.

* What's going to change? The default value of `allow_empty` used to be True. It will change to False.

* How to update your code? If a glob fails, use `allow_empty = True` to keep the old behavior.

In Bazel 8.0, the flag `--incompatible_disallow_empty_glob` will default to `true`. This means that the `allow_empty` argument defaults to `False` on all `glob` calls.

**Comment by Iulia-Soimaru** [comment]
It seems like `glob` function has a bug when using `allow_empty = True`, the [documentation](https://docs.bazel.build/versions/2.2.0/be/functions.html#glob) for bazel v 2.2.0 has this api for the function:
```bazel
glob(include, exclude=[], exclude_directories=1, allow_empty=True)
```

when I try to apply it in the `filegroup` as
```bazel
filegroup(
    name = "icons_svg",
    srcs = glob(
        [
            "src/icons/*.svg",
        ],
        allow_empty = True,
    ),
)
```

and then use this filegroup in another bazel rule
```bazel
ts_node(
    name = "icons_gen",
    args = [
        "src/scripts/myScript.ts",
        # more ts-node cli arguments ...
        "$(rootpaths :icons_svg)",
    ],
    data = [
        ":icons_svg",
    ],
)
```
I am getting an error
```bash
in args attribute of nodejs_binary rule //:icons_gen: label '//:src/icons' in $(location) expression is not a declared prerequisite of this rule. Since this rule was created by the macro 'ts_node', the error might have been caused by the macro implementation
```

**Comment by alandonovan** [comment]
Status: still awaiting cleanup and flag flip.

**Comment by martis42** [comment]
This is labeled with `breaking-change-5.0`, which I interpret as the intention to flip this with Bazel 5.0.0.
The flag is however still inactive by default: 
https://github.com/bazelbuild/bazel/blob/5.0.0/src/main/java/com/google/devtools/build/lib/packages/semantics/BuildLanguageOptions.java#L355

Please update the issue status.

**Comment by meteorcloudy** [comment]
<img width="1153" alt="image" src="https://user-images.githubusercontent.com/4171702/190151600-a26eea86-81d8-4c67-8d1d-d25b1cd9fdf2.png">
https://buildkite.com/bazel/bazelisk-plus-incompatible-flags/builds/1253

This flag is breaking almost all downstream projects, I believe it's the same for the wider community. If no active plan on migrating the ecosystem, I'll remove "migration-ready" label for now.

**Comment by martis42** [comment]
@meteorcloudy 
I believe migrating this flag is a chicken and egg problem. Enabling this does not really solve a problem for users. It merely establishes a sane default for globbing.
Imho donwstream projects will most likely not adapt until there is an incentive.

What I am going for is, imho we should flip this flag soon (maybe Bazel 6.0?). This might be the gentle push for downstream projects to adapt to the flag. If they don't want to, setting the old behavior is easy enough.

**Comment by meteorcloudy** [comment]
> What I am going for is, imho we should flip this flag soon (maybe Bazel 6.0?). This might be the gentle push for downstream projects to adapt to the flag. If they don't want to, setting the old behavior via the incompatibility flag is easy enough.

No, this doesn't really work. You can see in the link I posted, flipping this flag will break almost all the downstream projects. Many of them are common dependencies in the ecosystem instead of just leaf projects. If your dependency doesn't build with this flag, you will be forced to unflip the flag in your bazelrc file even if you have migrated your own code.

So the correct way is to file issues to downstream projects, make sure common dependencies in the ecosystem are migrated first, then flip the flag. 

See our updated incompatible change process: https://bazel.build/contribute/breaking-changes?hl=en#update-repos

**Comment by limdor** [comment]
@bazel-io flag

**Comment by limdor** [comment]
Ok, probably I have to retract and rephrase, now most of the work is done for the Bazel project itself, if we want to fix the downstream projects there is still quite a lot left.

**Comment by meteorcloudy** [comment]
@limdor The flag wasn't tested in the pipeline because the "migration-ready" was removed. I added it again let's see https://github.com/bazelbuild/bazel/pull/15374 will fix the failures.

**Comment by meteorcloudy** [comment]
From https://buildkite.com/bazel/bazelisk-plus-incompatible-flags/builds/1253#01833aea-8848-4562-976c-916e87f33b6b, I can see at least TensorFlow is broken by this: 
```
(08:59:35) ERROR: Traceback (most recent call last):
	File "C:/b/bk-windows-gclq/bazel-downstream-projects/tensorflow/tensorflow/tools/docs/BUILD", line 154, column 16, in <toplevel>
		data = glob(["**/create_model.md"]),
Error in glob: glob pattern '**/create_model.md' didn't match anything, but allow_empty is set to False (the default value of allow_empty can be set with --incompatible_disallow_empty_glob).
```

**Comment by limdor** [comment]
> @limdor The flag wasn't tested in the pipeline because the "migration-ready" was removed. I added it again let's see #15374 will fix the failures.

Ok understood, I will take a look to the failures then. In general the fixes are quite trivial, at least to get the default behavior. The problem is getting reviews in timely manner.
The problem that I might be facing in downstream projects is that setting the allow_empty = True might not be well received and some projects might want to investigate why that is the case. However my rationale here is that first it should be set allow_empty to True and then investigate, with this way we can flip the default behavior and stop introducing cases where the glob is empty but the developer did not expect that.

I understand that the downstream projects are important but at some point there is the whole Bazel comunity that is suffering from not flipping this flag.

My proposal would be to try to get the PRs for the downstream projects where it sets allow_empty to true for the affected globs, and in the hypothetical case where they don't want to merge it, consider flipping it anyway. But somehow I'm sure we should be able to convince the maintainers of the downstream projects that setting the default explicitly is better than doing nothing.

**Comment by meteorcloudy** [comment]
> My proposal would be to try to get the PRs for the downstream projects where it sets allow_empty to true for the affected globs, and in the hypothetical case where they don't want to merge it, consider flipping it anyway.

Agree, however, there is no clear owner of this flag from the Bazel team right now, so it would be great if the community can help send those PRs. It will be useful to have a tracking list like https://github.com/bazelbuild/continuous-integration/issues/1404#issue-1341836598

**Comment by limdor** [comment]
> > My proposal would be to try to get the PRs for the downstream projects where it sets allow_empty to true for the affected globs, and in the hypothetical case where they don't want to merge it, consider flipping it anyway.
> 
> Agree, however, there is no clear owner of this flag from the Bazel team right now, so it would be great if the community can help send those PRs. It will be useful to have a tracking list like [bazelbuild/continuous-integration#1404 (comment)](https://github.com/bazelbuild/continuous-integration/issues/1404#issue-1341836598)

For the moment the tracking list is here https://github.com/bazelbuild/bazel/pull/15327#issue-1213646746

**Comment by meteorcloudy** [comment]
@limdor The list seems to track the progress of fixing Bazel itself? There are many downstream projects also broken, which should also be migrated first before flipping the flag: [https://buildkite.com/bazel/bazelisk-plus-incompatible-flags/builds/1253](https://buildkite.com/bazel/bazelisk-plus-incompatible-flags/builds/1253#01833aea-8848-4562-976c-916e87f33b6b)

**Comment by limdor** [comment]
> @limdor The list seems to track the progress of fixing Bazel itself? There are many downstream projects also broken, which should also be migrated first before flipping the flag: [https://buildkite.com/bazel/bazelisk-plus-incompatible-flags/builds/1253](https://buildkite.com/bazel/bazelisk-plus-incompatible-flags/builds/1253#01833aea-8848-4562-976c-916e87f33b6b)

Initially was, but I'm planning to extend it. Any chance to leave the migration ready flag? If not it's quite hard to track and I have to ask every time.

**Comment by meteorcloudy** [comment]
@limdor Sure, as long as someone is actively working on this, I'm happy to enable the test!

**Comment by limdor** [comment]
> @limdor Sure, as long as someone is actively working on this, I'm happy to enable the test!

Thanks! As active as can be doing it in my free time.

**Comment by meteorcloudy** [comment]
@limdor Thank you so much!

**Comment by limdor** [comment]
Update: All PRs in Bazel have been merged except the flag flip itself. The remaining part is the downstream projects that are tracked here https://github.com/bazelbuild/bazel/pull/15327

**Comment by keertk** [comment]
Hi all, wanted to check if we plan to flip this flag in 7.0? Let me know when this is ready for migration so we can check for downstream fixes needed. Thanks.

**Comment by brentleyjones** [comment]
`allow_empty = False` causing errors when any pattern is empty can result in unsafe globs by forcing people to use `allow_empty =True`. Take the following `glob` as an example:

```
glob([
   allow_empty = False,
    "*.strings",
    "*.stringsdict",
])
```

I want an error if we don't have any `.strings` or `.stringsdict` files. But I'm fine with having either. Currently this will error out if I don't have both. So I need to set `allow_empty = True`, but then I won't get errors when its empty.

**Comment by fmeum** [comment]
@brentleyjones Would this work for you?

```starlark
glob([
   allow_empty = True,
    "*.strings",
    "*.stringsdict",
]) or fail("glob is empty")
```

**Comment by brentleyjones** [comment]
Yes. Thank you.

**Comment by zachburg** [comment]
Would it be possible to add separate `allow_empty` attributes for patterns and the entire glob?

I want to use a specific patterns in https://github.com/abrisco/rules_helm/blob/main/helm/private/helm.bzl#L75 to match certain files, but I found I need to set `allow_empty = True`:

```
templates = native.glob(
            [
                "templates/*.yaml",
                "templates/*.yml",
                "templates/*.tpl",
                "templates/*.txt",
                "templates/tests/*.yaml",
                "templates/tests/*.yml",
            ],
            # Each pattern is individually evaluated, so allow_empty = True is necessary:
            allow_empty = True,
        )
```

The glob patterns are based on https://github.com/helm/helm/blob/a73c51ca08297fda17f40b3b11ff602e22893334/pkg/lint/rules/template.go#L208 and https://helm.sh/docs/topics/chart_tests/.

It's OK if no files match `templates/*.yml`, but it's not OK if the entire glob is empty.

I believe the specific issue I have here is from https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/packages/GlobCache.java#L272.

**Comment by fmeum** [comment]
You can achieve this by appending `or fail("...")`.

**Comment by meisterT** [comment]
This incompatible change has been completed: `--incompatible_disallow_empty_glob` was flipped to true by default in 538b1194b2687f97ad9b069b6b41a488d53a5a97 (PR #15327) for Bazel 8.
